libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/KotlinPlugin.kt

```diff
@@ -181,30 +178,37 @@ internal class Kotlin2JvmSourceSetProcessor(
 
     override fun doTargetSpecificProcessing() {
         ifKaptEnabled(project) {
-            Kapt3KotlinGradleSubplugin.createAptConfigurationIfNeeded(project, kotlinCompilation.compilationName)
+            Kapt3GradleSubplugin.createAptConfigurationIfNeeded(project, kotlinCompilation.compilationName)
         }
 
         ScriptingGradleSubplugin.configureForSourceSet(project, kotlinCompilation.compilationName)
 
-        // TODO: here, the tasks are always triggered for configuration; once the subplugins are able to work without task instances,
-        //       ensure that task configuration is properly avoided here;
         project.whenEvaluated {
-            val kotlinTaskInstance = kotlinTask.get()
-            val javaTask = javaSourceSet?.let { project.tasks.findByName(it.compileJavaTaskName) as JavaCompile }
-
             val subpluginEnvironment = SubpluginEnvironment.loadSubplugins(project, kotlinPluginVersion)
+
             val appliedPlugins = subpluginEnvironment.addSubpluginOptions(
-                project, kotlinTaskInstance, javaTask, null, null, kotlinCompilation
+                project,
+                kotlinCompilation,
+                javaSourceSet?.let { tasks.withType(AbstractCompile::class.java).named(it.compileJavaTaskName) }
             )
 
             appliedPlugins
-                .flatMap { it.getSubpluginKotlinTasks(project, kotlinTaskInstance) }
-                .forEach { plugin -> kotlinCompilation.allKotlinSourceSets.forEach { sourceSet -> plugin.source(sourceSet.kotlin) } }
+                .flatMap { it.getPluginKotlinTasks(kotlinCompilation) }
+                .forEach { pluginTaskProvider ->
+                    pluginTaskProvider.configure { pluginTask ->
+                        kotlinCompilation.allKotlinSourceSets.forEach { sourceSet -> pluginTask.source(sourceSet.kotlin) }
+                    }
+                }
 
-            javaTask?.let { configureJavaTask(kotlinTaskInstance, it, logger) }
+            kotlinTask.configure { kotlinTaskInstance ->
+                if (project.pluginManager.hasPlugin("java-library") && sourceSetName == SourceSet.MAIN_SOURCE_SET_NAME) {
+                    registerKotlinOutputForJavaLibrary(kotlinTaskInstance.destinationDir, kotlinTaskInstance)
+                }
+            }
 
-            if (project.pluginManager.hasPlugin("java-library") && sourceSetName == SourceSet.MAIN_SOURCE_SET_NAME) {
-                registerKotlinOutputForJavaLibrary(kotlinTaskInstance.destinationDir, kotlinTaskInstance)
+            javaSourceSet?.let { java ->
+                val javaTask = project.tasks.withType(AbstractCompile::class.java).named(java.compileJavaTaskName)
```

Comment:
You can use `named(java.lang.String, java.lang.Class)`

Reply:
Yes, indeed we can, but we couldn't at the moment I made this change. :) 
https://github.com/JetBrains/kotlin/pull/2892#discussion_r372210086